### DISCUSSION TITLE
iterator for  vertices at both ends forming an edge

### DISCUSSION THREAD

**Discussion by zinuok** [CGAL/cgal#6135] [Q&A] [answered]
## Issue Details

I'm using "CGAL/Triangulation_3.h" header to generate tetra 3D mesh from 3D points.
I succeed to generate 3D tetra mesh and I want to access each edge. To be precise, I want to get the coordinates of the vertex points that make up each edge.
However, I can't find any suitable example for these kinds of iterator

## Source Code

```
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Triangulation_3.h>

// type definition
typedef CGAL::Triangulation_3<K>      Triangulation_cgal;
typedef Triangulation_cgal::Point          Point_cgal;

// make 3D mesh from 3D points (total 6 points here)
std::list<Point_cgal> L;
L.push_front(Point_cgal(0,0,0));
L.push_front(Point_cgal(1,0,0));
L.push_front(Point_cgal(0,1,0));

L.push_front(Point_cgal(0,0,1));
L.push_front(Point_cgal(1,1,1));
L.push_front(Point_cgal(2,2,2));
Triangulation_cgal T(L.begin(), L.end());

// I want to access the end vertices of each edge
for ( auto it = T.edges_begin(); it != T.edges_end(); ++it ) {
    double x = it->[for what method to access to end vertices?]vertex()->point().x();
}

```

thank you 

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Linux (Ubuntu 18.04)
* Compiler: c++11
* Release or debug mode: Release
* Specific flags used (if any):
* CGAL version: 5.3
* Boost version: 
* Other libraries versions if used (Eigen, TBB, etc.):

**Comment by MaelRL** [comment, marked as answer]
An edge is a `Triple<Cell_handle, int, int>`. See [3D triangulations representations](https://doc.cgal.org/latest/TDS_3/index.html#title0) and more generally [the 3D triangulation data structure reference manual](https://doc.cgal.org/latest/TDS_3/group__PkgTDS3Ref.html) (e.g. https://doc.cgal.org/latest/TDS_3/classTriangulationDataStructure__3_1_1Cell.html will give you the API of the cell). Here, `it->first` will give you the cell, `it->first->vertex(index)` the vertex, `it->first->vertex(index)->point()` the point, etc.

If you are confused by the doc, you can also search the CGAL code base to see how functions are used in practice.
